docs/reinterpreting_effects.md

```diff
@@ -0,0 +1,359 @@
+      , responseClose' = ResponseClose (pure ())
+      }
+
+badJson :: HTTP.Request -> IO (HTTP.Response L.ByteString)
+badJson req = pure resp
+  where
+    resp = Response
+      { responseStatus = ok200
+      , responseVersion = http11
+      , responseHeaders = [("Content-Type", "application/json; charset=utf-8")]
+      , responseBody = "{}"
+      , responseCookieJar = mempty
+      , responseClose' = ResponseClose (pure ())
+      }
+```
+
+Let's update our `main` function and watch it in action:
+
+``` haskell
+main :: IO ()
+main = do
+  -- Should return JsonParseError
+  runMockHttp badJson catFactsRunner >>= handlePrint
+  -- Should return InvalidContentType
+  runMockHttp wrongContentType catFactsRunner >>= handlePrint
+```
+
+Which returns:
+
+``` haskell
+JsonParseError "Error in $: parsing [] failed, expected Array, but encountered Object"
+InvalidContentType "Just \"text/xml\""
+```
+
+With effects, we have fine-grained ways of testing slices of our API. All that's needed
+to turn an integration test into a unit test or vice versa is a different set of `Algebra`-implementing effect handlers!
+
+## Observing & altering effects
+
+Building new effect handling algebras can be a little bit verbose. In simpler situations, we may want to simply operate
+on an effect without having to implement a whole new `Algebra` instance. We still have yet to build a solution to tracking
+operational metrics (like request timings), so let's look at how to build a sort of "effect middleware" using `InterpretC`.
+
+`InterpretC` is an effect carrier that is intended for prototyping new effects that passes a callback function each
+occurence of the specified effect type that is called via `send`. One trick that can be useful is to intercept an effect,
+operate on the effect, and then re-`send` the effect. In other words, it's perfectly valid to have multiple handlers
+for the same effect type and dispatch to the ones higher in the effect stack! Let's use this approach to time and log
+our HTTP requests:
+
+``` haskell
+traceHttp
+  :: (Has Http sig m, MonadIO m)
+  => (forall s. Reifies s (Handler Http m) => InterpretC s Http m a)
+  -> m a
+traceHttp = runInterpret $ \r@(SendRequest req sendResp) -> do
+  startTime <- liftIO getCurrentTime
+  liftIO (putStr (B.unpack (HTTP.path req) ++ " ... "))
+  -- Pass the request on to something that actually knows how to respond.
+  send $ SendRequest req $ \resp -> do
+    -- Once the actual response is obtained,
+    -- we can capture the end time and status of the response.
+    endTime <- liftIO getCurrentTime
+    let timeSpent = endTime `diffUTCTime` startTime
+    liftIO $ putStrLn ("[" ++ show (statusCode $ HTTP.responseStatus resp) ++ "] took " ++ show timeSpent ++ "\n\n")
+    sendResp resp
+```
+
+Updating our `main` function once more:
+
+``` haskell
+main :: IO ()
+main = runHttp (traceHttp catFactsRunner) >>= handlePrint
+```
+
+Returns:
+
+```
+/facts/random ... [200] took 0.979107082s
```

Comment:
_Beautiful_. Wow.

Reply:
Thanks, I'm flattered! Although I'm actually not sure what part of this is merits this response 😄

Reply:
Layering on timing like this was a really sweet solution ✨